File: frontend/modules/thryft-video-export/ios/VideoExportError.swift
//
//  VideoExportError.swift
//  thryft-video-export
//
//  Error type mirroring the JS-side `VideoExportError` discriminated union.
//  `code` maps to the union's `type` field; the JS wrapper
//  (`toVideoExportError` in src/index.ts) parses the message prefix
//  `"{code}:{message}"` back into the union, so the tag must stay in sync.
//

import Foundation

enum VideoExportFailure: Error {
  case cancelled
  case unsupported(String)
  case renderFailed(String)
  case invalidInput(String)

  /// The `type` tag of the JS-side `VideoExportError` union.
  var code: String {
    switch self {
    case .cancelled: return "cancelled"
    case .unsupported: return "unsupported"
    case .renderFailed: return "render_failed"
    case .invalidInput: return "invalid_input"
    }
  }

  /// Message surface to JS. Prefixed with the code so the JS wrapper can
  /// recover the discriminated union without parsing localised strings.
  var bridgedMessage: String {
    switch self {
    case .cancelled:
      return "cancelled: Export was cancelled."
    case .unsupported(let detail):
      return "unsupported: \(detail)"
    case .renderFailed(let detail):
      return "render_failed: \(detail)"
    case .invalidInput(let detail):
      return "invalid_input: \(detail)"
    }
  }
}

extension VideoExportFailure: LocalizedError {
  var errorDescription: String? { bridgedMessage }
}

File: frontend/modules/thryft-video-export/android/src/main/java/com/margelo/nitro/videoexport/VideoExportError.kt
//
//  VideoExportError.kt
//  thryft-video-export
//
//  Error type mirroring the JS-side `VideoExportError` discriminated union.
//  `code` maps to the union's `type` field; the JS wrapper
//  (`toVideoExportError` in src/index.ts) parses the message prefix
//  `"{code}:{message}"` back into the union, so the tag must stay in sync.
//

package com.margelo.nitro.videoexport

class VideoExportException(
  val code: String,
  detail: String,
) : Exception("$code:$detail") {
  companion object {
    fun cancelled() = VideoExportException("cancelled", "Export was cancelled.")
    fun unsupported(detail: String) = VideoExportException("unsupported", detail)
    fun renderFailed(detail: String) = VideoExportException("render_failed", detail)
    fun invalidInput(detail: String) = VideoExportException("invalid_input", detail)
  }
}
